### PR TITLE
[X86] Support SM4 EVEX version intrinsics/instructions.

### PR DESCRIPTION
**Pull request by FreddyLeaf**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/671368

### CODE REVIEW

**Review by KanRobert — APPROVED** [review]
LGTM

File: llvm/docs/ReleaseNotes.md [hunk 1]
```diff
@@ -196,6 +196,8 @@ Changes to the X86 Backend
 
 * Support ISA of `AVX10.2-256` and `AVX10.2-512`.
 
+* Support ISA of `SM4(EVEX)`.
```
Comment (phoebewang) [inline llvm/docs/ReleaseNotes.md:199]: We should mainly add in Clang release note.

File: llvm/include/llvm/IR/IntrinsicsX86.td [hunk 2]
```diff
@@ -6109,6 +6109,12 @@ let TargetPrefix = "x86" in {
         DefaultAttrsIntrinsic<[llvm_v8i32_ty],
         [llvm_v8i32_ty, llvm_v8i32_ty],
         [IntrNoMem]>;
+def int_x86_vsm4key4512 : ClangBuiltin<"__builtin_ia32_vsm4key4512">,
+        Intrinsic<[llvm_v16i32_ty], [llvm_v16i32_ty, llvm_v16i32_ty],
+        [IntrNoMem]>;
+def int_x86_vsm4rnds4512 : ClangBuiltin<"__builtin_ia32_vsm4rnds4512">,
+        Intrinsic<[llvm_v16i32_ty], [llvm_v16i32_ty, llvm_v16i32_ty],
```
Comment (phoebewang) [inline llvm/include/llvm/IR/IntrinsicsX86.td:6116]: ditto.

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 3]
```diff
@@ -1647,3 +1647,23 @@ let isCodeGenOnly = 1, ForceDisassemble = 1, hasSideEffects = 0 in
...
+                "128", avx512vl_i32_info.info128.LdFrag,
+                avx512vl_i32_info.info128.MemOp>, EVEX_V128;
+    defm Z256 : SM4_Base<OpStr, avx512vl_i32_info.info256.RC,
+                "256", avx512vl_i32_info.info256.LdFrag,
+                avx512vl_i32_info.info256.MemOp>, EVEX_V256;
+  }
+  let Predicates = [HasSM4, HasAVX10_2_512] in
+    defm Z : SM4_Base<OpStr, avx512vl_i32_info.info512.RC,
+              "512", avx512vl_i32_info.info512.LdFrag,
+              avx512vl_i32_info.info512.MemOp>, EVEX_V512;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAVX10.td:1665]: Align within `<`

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 4]
```diff
@@ -1647,3 +1647,23 @@ let isCodeGenOnly = 1, ForceDisassemble = 1, hasSideEffects = 0 in
   def : InstAlias<"vmovw.s\t{$src, $dst|$dst, $src}",
                   (VMOVZPWILo2PWIZrr2 VR128X:$dst, VR128X:$src), 0>;
 }
+
+// SM4(EVEX)
+multiclass avx10_sm4_base<string OpStr> {
+  // SM4_Base is in X86InstrSSE.td.
+  let Predicates = [HasSM4, HasAVX10_2] in {
+    defm Z128 : SM4_Base<OpStr, avx512vl_i32_info.info128.RC,
+                "128", avx512vl_i32_info.info128.LdFrag,
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAVX10.td:1656]: `avx512vl_i32_info.info128.RC` -> `VR128X`
`vx512vl_i32_info.info128.LdFrag` -> `loadv4i32`
`avx512vl_i32_info.info128.MemOp` -> `i128mem`

File: llvm/test/CodeGen/X86/sm4-evex-intrinsics.ll [hunk 5]
```diff
@@ -0,0 +1,64 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -verify-machineinstrs -mtriple=x86_64-- --show-mc-encoding -mattr=+sm4,+avx10.2-512 | FileCheck %s
+; RUN: llc < %s -verify-machineinstrs -mtriple=i686-- --show-mc-encoding -mattr=+sm4,+avx10.2-512 | FileCheck %s
+
+define <4 x i32> @test_int_x86_vsm4key4128(<4 x i32> %A, <4 x i32> %B) {
+; CHECK-LABEL: test_int_x86_vsm4key4128:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    vsm4key4 %xmm1, %xmm0, %xmm0 # encoding: [0xc4,0xe2,0x7a,0xda,0xc1]
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/sm4-evex-intrinsics.ll:8]: I assume there's a comment about EVEX2VEX compress.

File: llvm/test/MC/Disassembler/X86/sm4-evex-32.txt [hunk 6]
```diff
@@ -0,0 +1,170 @@
+# RUN: llvm-mc --disassemble %s -triple=x86_64 | FileCheck %s --check-prefixes=ATT
```
Comment (phoebewang) [inline llvm/test/MC/Disassembler/X86/sm4-evex-32.txt:1]: The file name is mistaken with sm4-evex-64.txt

File: llvm/test/MC/Disassembler/X86/sm4-evex-64.txt [hunk 7]
```diff
@@ -0,0 +1,170 @@
+# RUN: llvm-mc --disassemble %s -triple=i386-unknown-unknown | FileCheck %s --check-prefixes=ATT
```
Comment (phoebewang) [inline llvm/test/MC/Disassembler/X86/sm4-evex-64.txt:1]: ditto.

File: llvm/test/MC/X86/sm4-evex-32-att.s [hunk 8]
```diff
@@ -0,0 +1,224 @@
+// RUN: llvm-mc -triple x86_64-unknown-unknown --show-encoding %s | FileCheck %s
```
Comment (phoebewang) [inline llvm/test/MC/X86/sm4-evex-32-att.s:1]: The same to below.

File: llvm/test/CodeGen/X86/sm4-evex-intrinsics.ll (hunk 5 above)
Comment (FreddyLeaf) [reply llvm/test/CodeGen/X86/sm4-evex-intrinsics.ll:8]: All machines supports SM4 EVEX version also supports SM4 VEX version. similar situations for AVX512F.

Comment (phoebewang) [reply llvm/test/CodeGen/X86/sm4-evex-intrinsics.ll:8]: So there will be a EVEX2VEX comment. The point is we need to select to EVEX instruction first and then optimize to VEX.

Comment (FreddyLeaf) [reply llvm/test/CodeGen/X86/sm4-evex-intrinsics.ll]: [39513cd](https://github.com/llvm/llvm-project/pull/113402/commits/39513cdd193f3dc48e7df635b50c70ea51f2adef)

File: llvm/test/MC/X86/sm4-evex-32-att.s (hunk 8 above)
Comment (FreddyLeaf) [reply llvm/test/MC/X86/sm4-evex-32-att.s]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: llvm/test/MC/Disassembler/X86/sm4-evex-64.txt (hunk 7 above)
Comment (FreddyLeaf) [reply llvm/test/MC/Disassembler/X86/sm4-evex-64.txt]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: llvm/test/MC/Disassembler/X86/sm4-evex-32.txt (hunk 6 above)
Comment (FreddyLeaf) [reply llvm/test/MC/Disassembler/X86/sm4-evex-32.txt]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: llvm/lib/Target/X86/X86InstrAVX10.td (hunk 4 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: llvm/lib/Target/X86/X86InstrAVX10.td (hunk 3 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: llvm/include/llvm/IR/IntrinsicsX86.td (hunk 2 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/IR/IntrinsicsX86.td]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: llvm/include/llvm/IR/IntrinsicsX86.td [hunk 9]
```diff
@@ -6109,6 +6109,12 @@ let TargetPrefix = "x86" in {
         DefaultAttrsIntrinsic<[llvm_v8i32_ty],
         [llvm_v8i32_ty, llvm_v8i32_ty],
         [IntrNoMem]>;
+def int_x86_vsm4key4512 : ClangBuiltin<"__builtin_ia32_vsm4key4512">,
+        Intrinsic<[llvm_v16i32_ty], [llvm_v16i32_ty, llvm_v16i32_ty],
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/include/llvm/IR/IntrinsicsX86.td]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: llvm/docs/ReleaseNotes.md (hunk 1 above)
Comment (FreddyLeaf) [reply llvm/docs/ReleaseNotes.md:199]: [23d4bfc](https://github.com/llvm/llvm-project/pull/113402/commits/23d4bfcabdca8968cba600e2fc0b440121990ef5)

File: clang/lib/Headers/immintrin.h [hunk 10]
```diff
@@ -665,6 +665,9 @@ _storebe_i64(void * __P, long long __D) {
 #include <avx10_2_512niintrin.h>
 #include <avx10_2_512satcvtdsintrin.h>
 #include <avx10_2_512satcvtintrin.h>
+#if (defined(__SM4__))
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/Headers/immintrin.h:668]: Oh, we should use
```
#if !defined(__SCE__) || __has_feature(modules) ||                             \
    (defined(__AVX10_2_512__) && defined(__SM4__))
#include <sm4evexintrin.h>
#endif
```

[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/immintrin.h]: [c0a33ab](https://github.com/llvm/llvm-project/pull/113402/commits/c0a33abe804bafe11e9dcecb092b924bcd08cdb7)

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**